
i.MXL MMC controller has a built-in card detection. This patch makes it
available for the one who doesn't want to sacrifice a GPIO to do that.

Detection relies on SD_DAT3 pin being pulled high when a card is
inserted. So chip's internal pull-down resistor is activated to have
this pin low by default (recommended by MMC specs in any case).

Note: requires to replace IO_ADDRESS(IMX_MMC_BASE) with host->base
      to be clean. (But how ?)

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>
Signed-off-by: Eric Jarrige <jorasse@armadeus.org>
 
Index: linux-2.6.29/drivers/mmc/host/imxmmc.c
===================================================================
--- linux-2.6.29.orig/drivers/mmc/host/imxmmc.c	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29/drivers/mmc/host/imxmmc.c	2009-04-30 23:12:36.000000000 +0200
@@ -911,6 +911,29 @@
 	.get_ro		= imxmci_get_ro,
 };
 
+/* Controller built-in card detection */
+static int imxmci_card_present(struct device *dev)
+{
+	static int current_state = 1;
+	static int nb_trials = 3;
+	/* struct imxmci_host *host = how to get it from dev ?? */
+	u16 status = readw(IO_ADDRESS(IMX_MMC_BASE) + MMC_REG_STATUS);
+	int new_state = (status & STATUS_CARD_PRESENCE) ? 1 : 0;
+
+	/* Take state into account only if data are not being transmitted */
+	if (!(status & STATUS_CARD_BUS_CLK_RUN)
+			&& (new_state != current_state)) {
+		if (0 == --nb_trials) {
+			current_state = new_state;
+			nb_trials = 3;
+		}
+	} else {
+		nb_trials = 3;
+	}
+
+	return current_state;
+}
+
 static void imxmci_check_status(unsigned long data)
 {
 	struct imxmci_host *host = (struct imxmci_host *)data;
@@ -1043,6 +1066,10 @@
 	if (ret)
 		goto out;
 
+	/* If no way of detecting card is given, use built-in method */
+	if (!host->pdata->card_present)
+		host->pdata->card_present = imxmci_card_present;
+
 	if (host->pdata && host->pdata->card_present)
 		host->present = host->pdata->card_present(mmc_dev(mmc));
 	else	/* if there is no way to detect assume that card is present */
Index: linux-2.6.29/arch/arm/mach-imx/include/mach/imx-regs.h
===================================================================
--- linux-2.6.29.orig/arch/arm/mach-imx/include/mach/imx-regs.h	2009-04-30 23:09:11.000000000 +0200
+++ linux-2.6.29/arch/arm/mach-imx/include/mach/imx-regs.h	2009-04-30 23:11:18.000000000 +0200
@@ -172,7 +172,7 @@
 #define PB9_AF_MS_PI1        ( GPIO_PORTB | GPIO_AF | 9 )
 #define PB10_PF_SD_DAT2      ( GPIO_PORTB | GPIO_PF | GPIO_PUEN  | 10 )
 #define PB10_AF_MS_SCLKI     ( GPIO_PORTB | GPIO_AF | 10 )
-#define PB11_PF_SD_DAT3      ( GPIO_PORTB | GPIO_PF | 11 )
+#define PB11_PF_SD_DAT3      ( GPIO_PORTB | GPIO_PF | GPIO_PUEN | 11 )
 #define PB11_AF_MS_SDIO      ( GPIO_PORTB | GPIO_AF | 11 )
 #define PB12_PF_SD_CLK       ( GPIO_PORTB | GPIO_PF | 12 )
 #define PB12_AF_MS_SCLK0     ( GPIO_PORTB | GPIO_AF | 12 )
